### ISSUE TITLE
Adding Regular SMS Support

### ISSUE THREAD

**Issue by applerebel** [shusain93/Andromeda-iMessage#14]
Is it possible to add SMS support in this app that way we can use one app for both SMS and iMessage like on the iPhone. I imagine it would have to check against previous iMessage conversations to see if the user is an iMessage user or not.

**Comment by ezhes** [comment]
Possible, yes, and it’d be totally sweet. I just sort of bought an iPhone 8 so the future of this project is uncertain. I’ve still got my pixel and I do tend to switch back and forth all the time so maybe I’ll get around to it. What I did was just keep SMS and Andromeda in one folder so that I can send photos to people over sms as needed

**Comment by applerebel** [comment]
Thanks for the quick response. I think I can help in regards to the image sending/receiving and the SMS. I also have an iPhone and Pixel (one is personal and one is business). If you are still interested in the project, I would be glad to help. I'm pretty fluent in Applescript and web development. However, I do not know how to code android apps. If you would be open to helping me with the Android App side of the development I would enjoy working with you to further the project.

**Comment by ezhes** [comment]
Actually if you can figure out any way to take a file (stored on the desktop or whatever) and get it into the messaging field that would be great! [The AppleScript is here](https://github.com/shusain93/OSXMessageProxy/blob/master/MessageProxy/messageSender.applescript). The route you'll probably want to take is usingUIAutomation == true however some cursosry research shows that the "fast messaging" (aka AppleScript commands and not UI Automation) supports the `send file screenShotFile to buddy <xxxxx>` format. 

If you did a pull request I'd feel obligated to integrate it into the Android and server apps lol, that'd be great!

**Comment by applerebel** [comment]
Do you have a personal email address where I may contact you directly?

**Comment by ezhes** [comment]
Sorry, I prefer everything out in the open. If you want to discuss this feature specifically you should create an issue over on the server repo. I don’t mind having an issue with three replies to it, GitHub manages that just fine.

**Comment by applerebel** [comment]
Tried to create a thread on the server. It seems it got deleted would you rather just use this thread? Also, I have a working PHP, HTML5 example using MAMP for the server. 
Things to do:
Add group message sending
Auto refresh messaging page
Location using (LOC.VCF files)
Notifications
Importing it into the ionic framework (for mobile app)
Emoji text size (detect and make emoji's bigger than regular text)

**Comment by ezhes** [comment]
Oh that’s awesome, so it’s a web client? When you get a good working example I’ll include it as an option on the README for other people. 

For the auto refreshing message pane: there’s a web socket running on I think 8376? It sends updated data packets as necessary so you can just watch the raw socket. I’m hoping to someday go to SSL/WebScokets SSL but it’s a ways off.

**Comment by applerebel** [comment]
Yep its a web client right now. Going to convert it to JSON when i get done with the basics. SSL will be easy if we switch off of Xcode based Webserver. Then we can use Let’s Encrypt (Free ssl). 

Do you think you can take JSON and turn that into an android app? If not we can use JSON with the Ionic framework. Correct me if I am wrong but we could have the app do JSON requests every 5 seconds for updating the message log when the app is open. When it is closed it won’t really matter since they will get a notification and on opening the app it would auto refresh.

**Comment by ezhes** [comment]
I decided to do socket updates because the whole thing is a lot of data depending on the number of conversations you have. It’s not very responsive and very “loud” on the network. Not sure what you’re on about with JSON, the Android app I built already uses it and really it’s the only way to get data out of the server.

Also, the web server is never going to be separated from the server project in Xcode. I was just planning on switching network libraries to a new one which supports SSL. The android app would literally just need an s added in front of the network calls and we’d be good.

Also, I’m using sockets for new individual messages because you don’t want to have to wait for the whole conversation to download again every time a person responds in a chat, and plus that delay might get annoying. I’d really recommend the socket

**Comment by applerebel** [comment]
Does your current coding work with custom named group chats?

**Comment by ezhes** [comment]
Yes. It's sort of confusing and I'm not sure how I set it up but somewhere in each JSON conversation there is a flag that's something like "hasCustomName" which means that the group chat has a specific name set and isn't just the comma separated list of people.

**Comment by applerebel** [comment]
Currently trying to make a simpler database to use by converting the Apple .db to MSQL. 

With the Apple database you keep having to go to the *_join tables. with the new database:

every message includes the chat_id
chat includes chat_id, chat identifier (who to send it to), type of chat, and display name (for custom group names)

This should make the looks ups faster and also has some added benefits. Such as, pushing the messages to an external PHP, MYSQL server for cases in which the computer that retrieves the imessages goes offline or in cases where that computer can't have open ports to the outside (Home ISP's). 

JSON request should be faster too with the new method.

**Comment by ezhes** [comment]
Were you having trouble with performance of the database? I had 5000 rows returns in about 3ms on my 2010 macbook air. The greatest bottleneck is sending the entire conversation/conversation list over the network. On my local wifi and on LTE it's not an issue however once I started getting way out into the mountains on 3G/Edge it was painful.

What would be insanely cool would be synchronizing the database to the device. Pull the db from the server once and just do delta updates from the pushes we receive. I was thinking about doing that with firebase (before I open sourced the project) but it quickly became way too complex). Maybe we could do a block level delta sync of the sqlite database that apple provides? That'd be cool.

When you really get down into it the join system makes a ton of sense, especially when you're grabbing multiple attachments and various weird metadata things. It's not all necessary NOW and so it's much cheaper to just keep the JOIN ID in memory and pick it up later. If you're doing the offloading to an external server that might make sense though? Also yeah I with how hacky this project is I never intended for people who can't port forward to use the system. There are some interesting services like ngork which can offer a reverse tunnel which might be cool for those people.

**Comment by applerebel** [comment]
I started the sync process sorta I need your help on the android app but my new database allows for storing of sync times. So you could add in your code something like check database table “data” for last synced message_id and current last message. If current last message is greater than last synced message_id run the update json request.

**Comment by applerebel** [comment]
Also, in response to your first half of the message. I wasn’t having problems but nesting numerous sql request can be a pain so I decided to make an easier to use database. Also the nesting when using a mobile network may slow down the request. Just trying to streamline server request as must as possible and make them more efficient. 

The best way for the user to use this service would be have the iMessage gathering computer be on a cellular network (WiFi hotspot) and then send the messages to a godaddy or hostgator hosting account since they would always be available and have speeds higher than a home network. 

Basically all the home server will do is push the message to the cloud server (godaddy, hostgator)

**Comment by ezhes** [comment]
I'm not exactly sure what you need. Since you're messing around with converting it, could you give me an example query with pseudo parameters? Do you more or less just want a `messagesForChat:<id> afterMessageID<id> limit<count>` API call? If you have two servers which are continuously available, the best thing to do would be hack the `checkForNewMessages` call in the DatabaseConnector file and replace the `sendNotification` function with something which sends the new message to your server for distribution. The issue with exactly what you're asking for is that the server would then need to keep track of the client's states more than ephemerally. Since this server is primarily designed to be server to user, not server-server, I built all the calls based on what a real user would be wanting to do.

Also, the network type doesn't matter when making SQL requests because really it's all abstracted away into little HTTP calls and JSON. Also, are you more concerned with uptime or speed? Is your home network slower than your phone's?

If you can open ports on your local home network, you could have your external server just connect to the socket and then use that for fast data transfer. 


In my testing the greatest issue was just the amount of data that had to be sent to the device. Despite pulling all the data from the server, Andromeda actually launches and is ready faster on my Pixel than on my iPhone 6 on a cold app start. We need to get caching somehow on endpoints to make it any faster. I'm not too well versed obviously in database replication. Maybe using end to end encryption using the API keys I could use Firebase (which is already part of the app for a beta test in push notifications and analytics). The limits for free firebase https://firebase.google.com/docs/database/usage/limits are huge and not an issue. Just slap AES on the any info other than maybe row or JOIN ids and privacy is good.

**Comment by ezhes** [comment]
Oh also firebase would solve the issue with push notifications being sent through IFTTT, the need for port forwarding, the insecure/SSL-less API thanks to cloud messaging.... Hmmmm.. worth looking into. Only issue would be attachments, those would likely still need to come in over the user's server since you can send massive files over iMessage.

**Comment by applerebel** [comment]
I haven't fully built the new PHP API yet. However could you look into how to read and send MMS and SMS on an android phone so that users can have one app for both regular texts and iMessage?

**Comment by ezhes** [comment]
I've already done that awhile ago. Building an SMS app is just really complicated and I honestly don't care enough. As I said in the first comment, it's a huge amount of effort for IMO a little bit of convenience. The goal of this project was to provide iMessage for Android phones in a usable way, not to make a product that's worthy of someone buying it.

If you have experience with C# look at https://play.google.com/store/apps/details?id=com.disa&hl=en. It supports plugins so you could write an iMessage plugin that works with this server. I don't remember why exactly I gave up on this lead so maybe it'd be worth looking into for you. I believe the issue I had was that I didn't want to deal with making a new database format for storing data on the device (which is required with Disa) but since you've already done that I'd give it a shot.

**Comment by applerebel** [comment]
Disa seems like it would make this whole process easier and we would mainly have to build the server/API side. Seems like it would fix the SMS/MMS if people want a unified messaging app like on the iPhone.

**Comment by applerebel** [comment]
If we use firebase with the php to put the data on it, I'll try to get working on this. Ill send you an example json when I am ready.

**Comment by ezhes** [comment]
With Firebase there wouldn't be a need for another server because all the data would be continuously available on Google's servers, synced, cached, and SUPER fast. What do you see a PHP script as being necessary here for?

**Comment by applerebel** [comment]
In order to push messages from the computer running iMessage to the firebase. Instead of using Xcode.

**Comment by applerebel** [comment]
Also does firebase do notifications?

**Comment by applerebel** [comment]
Just read over the Firebase Dev requirements. It appears it only uses HTTPS and SSL which is a good thing but would require new users to buy an SSL and probably hosting anyway.

**Comment by applerebel** [comment]
Hey shusain, I have a little bit more to do for my web based imessage: image sending, tidying everything up, and getting contact names.

**Comment by ezhes** [comment]
Firebase has OS X frameworks. It would all run from the single OS X server. No extra server would be required. I’ll have some time this weekend so I’ll look at Firebase integration

**Comment by applerebel** [comment]
How are you going to get around the necessity of SSL with HTTPS?

**Comment by applerebel** [comment]
I don't think mac supports the open source SSL's quite yet

**Comment by ezhes** [comment]
LOL yes of course macs support SSL. You really can't do anything without SSL.

**Comment by applerebel** [comment]
No, I understand macs can use SSL, but that’s going to require people to buy an ssl cert for the project right?

**Comment by ezhes** [comment]
Not for Firebase transactions, but for the media transfer that would be necessary if they want automatically trusted certificates. Self-signed certificates can be 'bonded' on client devices so after the first accepted connection (which the user is warned about) they will automatically accept **that and only that** certificate without warning.

**Comment by applerebel** [comment]
Firebase solves the problem of auto sync but does it help with the attachments?

**Comment by ezhes** [comment]
How much do you know about certificates and encryption? 

Self-signed certificates are the solution to free but encrypted communications, assuming the user trusts the network they initially communicate with the server over. Firebase, as I've said before, will not handle attachments. The existing server, using these self-signed certs, will send attachments to client devices. Attachments are so unmission critical I don't think they need to be mirrored. If I do something stupid and hacky like using Google Drive and Google Sheets (LOL but...) then files could be stored in there too. hahaha might actually look at that.

**Comment by applerebel** [comment]
I know a good amount about certificates and encryption on Linux and Windows systems not much on mac. Self signed certificates might not always say not trusted on some programs. Not sure if Firebase requires a trusted SSL or not.

**Comment by ezhes** [comment]
You've misunderstood the architecture. 

It's a dual system. Firebase is Google. It's under Google's certificates and will use Google/other auth. Firebase will handle all the messages and conversations in a constant sync database. Then there's the local server. This local Mac server will use its own self-signed certificate (or you can drop in your own certificate). The certificate will be 'trusted' ***by the client Android device***. This is just to encrypt the media in transit and protect the API token.

**Comment by applerebel** [comment]
Oh, I’m talking about the trusted status when you are pushing from the local server to the Firebase. I have had problems in the past pushing from a local server to Microsoft Azure’s database when using an untrusted SSL.

**Comment by ezhes** [comment]
I mentioned a bit back that in this relationship the server is acting like a client when communicating with firebase. As a result, the server doesn't need an ssl cert of its own because it's communicating to Google. Google has valid certificates obviously.

**Comment by applerebel** [comment]
I see so firebase is automatically gathering the information from the local server. Not the local server pushing to the firebase server.

**Comment by ConProgramming** [comment]
Intergrating sms and optimizing for Chromebooks as well would be a dream... I have android app and firebase experience if I can help at all

**Comment by ezhes** [comment]
If you have experience and have questions let me know. I bought an iPhone
so it’s sort of abandoned for the time being since I’ve got some cool iOS
projects going but I can answer anything you need.

**Comment by ConProgramming** [comment]
Yeah I can work on adding sms support and maybe even syncing with firebase
